### PR TITLE
Scissor support

### PR DESCRIPTION
**Pull request by PabloPL**
This commit adds scissor support.

- for non zero scissor we just add plbu cmd with scissor (taken from lima-ng).
- for zero scissor we skip building plbu cmd after header and don't buld vs cmds (in previous patch i was trying to skip drawing in this case but then we would see state/image like in previous frame).
This behaviour has been found by comparing binary driver memory dump with and without zero scissor.

It was tested with following sample https://github.com/PabloPL/gfx/tree/scissor-test/gbm-surface-scissor
It's just a copy of gbm-surface-move where it should display following data:
- frame0 - normal image
- frame1 - non zero scissor
- frame2 - non zero scissor
- frame3 - zero scissor
- frame4 - normal image
Where non zero scissor is (0, 0, half of screen, half of screen), so result is top left quarter of screen.

Also during testing it's good to change this order, so first is zero scissor and later non zero, with previous implementation (where we're just skipping all in draw_vbo) this was not working (displaying the same image as in previous frame).

### CODE REVIEW

File: src/gallium/drivers/lima/lima_draw.c [hunk 1]
```diff
@@ -353,7 +353,7 @@ lima_pack_plbu_cmd(struct lima_context *ctx, const struct pipe_draw_info *info)
       ctx->buffer_state[lima_ctx_buff_pp_plb_rsw].size;
    plbu_cmd[i++] = 0x80000000 | (gl_position_va >> 4); /* RSW_VERTEX_ARRAY */
 
-   if (ctx->dirty & LIMA_CONTEXT_DIRTY_SCISSOR) {
+   if (ctx->rasterizer->base.scissor) {
```
Comment (yuq) [inline src/gallium/drivers/lima/lima_draw.c]: This makes every draw PLBU cmd has the scissor part if scissor is enabled. It seems to me that we just need the first draw that enabled the scissor to have this part and latter draw need not it if scissor is not dirty. Does your binary dump show the same thing?

**Review by yuq — COMMENTED** [review]
The first two patch can be merged to one.

**Comment by yuq** [comment]
So with the following order:
frame0: full image
frame1: zero scissor
what will show in frame1 with the current implementation, a black screen or the frame0 content?

BTW. what about the glclear issue you mentioned before?

**Comment by PabloPL** [comment]
> what will show in frame1 with the current implementation, a black screen or the frame0 content?

It'll be empty screen (because glClear is set to 0,0,0,0. If it would be set to for example green, it would be green). I got exactly the same result when running this code on my intel gpu.

**Comment by yuq** [comment]
But I remember that you did some experiment that the glClear will also be scissored so frame1 will be the same as frame0. It can also be find in the glClear spec that:
The scissor box bounds the cleared region.

Oh, seems it relate to the order of calling glClear and glScissor. In your gbm-surface-scissor, the glClear happens before glScissor, so the frame1 should be empty (glClear not scissored), but if glClear happens after glScissor, frame1 will be the same as frame0.

**Comment by PabloPL** [comment]
About glclear issue.
In some piglit tests, where i was looking for test code (for example https://github.com/mesa3d/piglit/blob/master/tests/spec/gl-1.0/scissor-clear.c#L45), they're not calling swapbuffers/flush. In this case i was getting info that test failed and after that saw on console/got breakpoint hit that lima_flush got called and job submited to kernel. Sometimes it wasn't even submited.

I'm rebuilding mesa (on my h3 device, so it takes some time..), but will check this again.
Maybe in this case it shoud be some kind of waiting for gpu to finish processing? I have somewhere fence support, based on vc4 code, using lima_submit_wait).

Edit 1:
After changing order (so scissor is before clear) there is error with clear (it is not clipped to scissor region), but this could be handled when working on clear support (also mentioned on wiki).

**Review by yuq — APPROVED** [review]
Looks good for me. Let's address the TODO after the lima_draw refine.
